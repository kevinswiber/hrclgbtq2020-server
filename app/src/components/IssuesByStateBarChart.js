import React, { useState } from 'react';
import * as d3 from 'd3';
import { AxisDomain, Tick, Orientation, TickLine, TickText } from './d3/Axis';
import { bar, issue } from './IssuesByStateBarChart.module.css'

export const IssuesByStateBarChart = ({ data }) => {
  const margin = { top: 30, right: 60, bottom: 10, left: 60 };
  const barHeight = 25;
  const height = Math.ceil((data.length + 0.1) * barHeight) + margin.top + margin.bottom;
  const width = 900

  const x = d3.scaleLinear()
    .domain([-6, 6])
    .rangeRound([margin.left, width - margin.right]);

  const y = d3.scaleBand()
    .domain(d3.range(data.length))
    .rangeRound([margin.top, height - margin.bottom])
    .padding(0.1);

  const yTickFormat = (i) => data[i].name;

  const xTickLabels = ['Worse', 'Better'];
  const xTickFormat = (i) => xTickLabels[x.domain().indexOf(i)];

  return (
    <div id="issues-by-state-chart">
      <svg viewBox={`0,0,${width},${height}`}>
        <g key="bars">
          {data.map((d, i) => {
            let width = Math.abs(x(d.value) - x(0));
            if (width === 0) {
              width = 2;
            }

            return (
              <g key={d.kind}>
                <rect
                  className={bar}
                  fill={d3.schemeSet1[d.value >= 0 ? 1 : 0]}
                  x={x(Math.min(d.value, 0))}
                  y={y(i)}
                  width={width}
                  style={{ width: `${width}` }}
                  height={y.bandwidth()} />
                <title>{d.policy}</title>
              </g>
            );
          })}
        </g>

        <g
          transform={`translate(${x(0)},0)`}
          fill="none"
          fontSize="12"
          fontFamily="sans-serif"
          textAnchor="end">
          <AxisDomain orient={Orientation.LEFT} tickSize="0" scale={y} />
          {y.domain().map((d, i) => {
            const orient = Orientation.LEFT;
            const textAttrs = {};
            if (data[i].value <= 0) {
              textAttrs.textAnchor = 'start';
              textAttrs.x = 6;
            }

            return (
              <Tick
                orient={orient} scale={y}
                value={d}
                key={data[i].kind}
                line={
                  <TickLine orient={orient} tickSize="0" />
                }
                text={
                  <TickText
                    orient={orient}
                    tickSize="0"
                    tickPadding="5"
                    className={issue}
                    value={d}
                    scale={y}
                    tickFormat={yTickFormat}
                    {...textAttrs} />
                }
              />
            );
          })}
        </g>

        <g
          transform={`translate(0, ${margin.top})`}
          fill="none"
          fontSize="10"
          fontFamily="sans-serif"
          textAnchor="middle">
          {x.domain().map((d, i) => {
            const orient = Orientation.TOP;
            return (
              <Tick
                orient={orient} scale={x}
                value={d}
                key={`x-axis-${d}`}
                line={
                  <TickLine orient={orient} />
                }
                text={
                  <TickText
                    orient={orient}
                    value={d}
                    scale={x}
                    tickFormat={xTickFormat} />
                }
              />
            );
          })}
        </g>
      </svg>
    </div >
  );
};