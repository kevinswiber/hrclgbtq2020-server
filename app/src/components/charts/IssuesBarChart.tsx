
import React from 'react';

export const IssuesBarChart = (): JSX.Element => <div/>;