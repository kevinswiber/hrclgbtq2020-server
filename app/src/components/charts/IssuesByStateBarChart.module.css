.bar {
  /*transition: x 500ms, width 500ms, fill 500ms;*/
  transition: all 500ms;
  pointer-events: all;
}

.issue {
  transition: transform 500ms;
}

.tooltip {
  transition: opacity 500ms, height 500ms;
  pointer-events: none;
  position: fixed;
  padding: 10px;
  width: 200px;
  background-color: white;
  box-shadow: 0.17rem 0.17rem 0.25rem rgba(0, 0, 0, 0.5);
}

#issue-table {
  margin: 10px;
}