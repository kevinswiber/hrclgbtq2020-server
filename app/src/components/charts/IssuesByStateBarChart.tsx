import React, { useState } from "react";
import * as d3 from "d3";
import {
  AxisDomain,
  BandTick,
  NumberTick,
  Orientation,
  TickLine,
  TickText,
} from "../d3/Axis";
import * as styles from "./IssuesByStateBarChart.module.css";
import { StateIssue } from "../../typings/types";

export const IssuesByStateBarChart = ({
  data,
}: {
  data: Array<StateIssue>;
}): React.ReactElement => {
  const [tooltip, setTooltip] = useState({
    style: {
      left: 0,
      top: 0,
      opacity: 0,
    },
    text: "",
  });

  const margin = { top: 30, right: 60, bottom: 10, left: 60 };
  const barHeight = 25;
  const height =
    Math.ceil((data.length + 0.1) * barHeight) + margin.top + margin.bottom;
  const width = 700;

  const x = d3
    .scaleLinear()
    .domain([-6, 6])
    .rangeRound([margin.left, width - margin.right]);

  const y = d3
    .scaleBand<number>()
    .domain(d3.range(data.length))
    .rangeRound([margin.top, height - margin.bottom])
    .padding(0.1);

  const yTickFormat = (i: number) => data[i].name;

  const xTickLabels = ["Anti-LGBTQ", "Pro-LGBTQ"];
  const xTickFormat = (i: number) => {
    return xTickLabels[
      x
        .domain()
        .map((d) => d.toString())
        .indexOf(i.toString())
    ];
  };

  const toggleTooltip = (
    text: string,
    options: { hide: boolean } = { hide: false }
  ) => {
    return () => {
      const opacity = options.hide ? 0 : 1;
      setTooltip((prevTooltip) => ({
        ...prevTooltip,
        style: { ...prevTooltip.style, opacity },
        text,
      }));
    };
  };

  const placeTooltip = (ev) => {
    const left = ev.clientX + 20;
    const top = ev.clientY;
    setTooltip((prevTooltip) => ({
      ...prevTooltip,
      style: { ...prevTooltip.style, left, top },
    }));
  };

  return (
    <div id="issues-by-state-chart">
      <div id="tooltip" className={styles.tooltip} style={tooltip.style}>
        {tooltip.text}
      </div>
      <svg viewBox={`0,0,${width},${height}`}>
        <g key="bars">
          {data.map((d, i) => {
            let width = Math.abs(x(d.value) - x(0));
            if (width === 0) {
              width = 2;
            }

            return (
              <g key={d.kind}>
                <rect
                  className={styles.bar}
                  style={{ width, transitionDelay: `${i * 20}ms` }}
                  onMouseOver={toggleTooltip(d.policy)}
                  onMouseMove={placeTooltip}
                  onMouseOut={toggleTooltip(d.policy, { hide: true })}
                  fill={d3.schemeSet1[d.value >= 0 ? 1 : 0]}
                  x={x(Math.min(d.value, 0))}
                  y={y(i)}
                  width={width}
                  height={y.bandwidth()}
                />
                {/*<title>{d.policy}</title>*/}
              </g>
            );
          })}
        </g>

        <g
          transform={`translate(${x(0)},0)`}
          fill="none"
          fontSize="12"
          fontFamily="sans-serif"
          textAnchor="end"
        >
          <AxisDomain
            orient={Orientation.LEFT}
            tickSize={0}
            range={y.range()}
          />
          {y.domain().map((d, i) => {
            const orient = Orientation.LEFT;
            const textAttrs: {
              textAnchor?: string;
              x?: number;
              className: string;
            } = {
              className: styles.issue,
            };
            if (data[i].value <= 0) {
              textAttrs.textAnchor = "start";
              textAttrs.x = 6;
            }

            return (
              <BandTick
                orient={orient}
                d3Scale={y}
                value={d}
                key={data[i].kind}
                line={<TickLine orient={orient} tickSize={0} />}
                text={
                  <TickText
                    orient={orient}
                    tickSize={0}
                    tickPadding={5}
                    value={d}
                    tickFormat={yTickFormat}
                    {...textAttrs}
                  />
                }
              />
            );
          })}
        </g>

        <g
          transform={`translate(0, ${margin.top})`}
          fill="none"
          fontSize="10"
          fontFamily="sans-serif"
          textAnchor="middle"
        >
          {x.domain().map((d) => {
            const orient = Orientation.TOP;
            return (
              <NumberTick
                orient={orient}
                d3Scale={x}
                value={d}
                key={`x-axis-${d}`}
                line={<TickLine orient={orient} />}
                text={
                  <TickText
                    orient={orient}
                    value={d}
                    tickFormat={xTickFormat}
                  />
                }
              />
            );
          })}
        </g>
      </svg>
    </div>
  );
};
