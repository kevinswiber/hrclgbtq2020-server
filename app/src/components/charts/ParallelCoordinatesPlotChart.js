import * as d3 from 'd3';
import React from 'react';
import { AxisDomain, Orientation } from '../d3/Axis';

export const ParallelCoordinatesPlotChart = ({ issues }) => {
  const ys = {};
  for (const issue of issues) {
    const domain = [];
    for (const state of issue.states) {
      if (domain.indexOf(state.value) === -1) {
        domain.push(state.value)
      }
    }
    ys[issue.name] = domain;
  }

  for (const [_k, v] of Object.entries(ys)) {
    v.sort(d3.ascending);
  }

  const margin = { top: 30, right: 60, bottom: 10, left: 60 };
  const spacer = 100;
  const height = 900;
  const width = Object.keys(ys).length * spacer + margin.left + margin.right;

  const xDomain = Object.keys(ys);
  const x = d3.scaleBand()
    .domain(Object.keys(ys))
    .range(Object.keys(ys).length * spacer);

  const [keys, values] = Object.entries(ys);
  return (
    <div id="parallel-chart">
      <svg viewBox={`0,0,${width},${height}`}>
        <g
          transform={`translate(${x(0)},${margin.top})`}
          fill="none"
          fontSize="12"
          fontFamily="sans-serif"
          textAnchor="end">
        </g>
      </svg>
    </div >
  )
};