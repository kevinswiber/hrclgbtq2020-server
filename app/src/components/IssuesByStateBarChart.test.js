import React from 'react';
import TestRenderer from 'react-test-renderer';

import { IssuesByStateBarChart } from './charts/IssuesByStateBarChart';

describe('IssuesByStateBarChart', () => {
  it('renders correctly', () => {
    const data = [
      {
        kind: 'TEST_ISSUE',
        name: 'Test Issue',
        policy: 'A policy for a test issue',
        value: 3
      },
      {
        kind: 'TEST_AGAIN_ISSUE',
        name: 'Test Again Issue',
        policy: 'A policy for a test again issue',
        value: -2
      }
    ];
    const testRenderer = TestRenderer.create(<IssuesByStateBarChart data={data}/>);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});