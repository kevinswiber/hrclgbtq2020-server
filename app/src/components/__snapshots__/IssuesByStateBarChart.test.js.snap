// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesByStateBarChart renders correctly 1`] = `
<div
  id="issues-by-state-chart"
>
  <div
    id="tooltip"
    style={
      Object {
        "left": 0,
        "opacity": 0,
        "top": 0,
      }
    }
  >
    
  </div>
  <svg
    viewBox="0,0,700,93"
  >
    <g>
      <g>
        <rect
          fill="#377eb8"
          height={23}
          onMouseMove={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "transitionDelay": "0ms",
              "width": 145,
            }
          }
          width={145}
          x={350}
          y={33}
        />
      </g>
      <g>
        <rect
          fill="#e41a1c"
          height={23}
          onMouseMove={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "transitionDelay": "20ms",
              "width": 97,
            }
          }
          width={97}
          x={253}
          y={58}
        />
      </g>
    </g>
    <g
      fill="none"
      fontFamily="sans-serif"
      fontSize="12"
      textAnchor="end"
      transform="translate(350,0)"
    >
      <path
        className="domain"
        d="M0.5,30.5V83.5"
        stroke="currentColor"
      />
      <g
        className="tick"
        opacity="1"
        transform="translate(0,44.5)"
      >
        <line
          stroke="currentColor"
          x2={-0}
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x={-5}
        >
          Test Issue
        </text>
      </g>
      <g
        className="tick"
        opacity="1"
        transform="translate(0,69.5)"
      >
        <line
          stroke="currentColor"
          x2={-0}
        />
        <text
          dy="0.32em"
          fill="currentColor"
          textAnchor="start"
          x={6}
        >
          Test Again Issue
        </text>
      </g>
    </g>
    <g
      fill="none"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      transform="translate(0, 30)"
    >
      <g
        className="tick"
        opacity="1"
        transform="translate(60.5,0)"
      >
        <line
          stroke="currentColor"
          y2={-6}
        />
        <text
          dy="0em"
          fill="currentColor"
          y={-9}
        >
          Anti-LGBTQ
        </text>
      </g>
      <g
        className="tick"
        opacity="1"
        transform="translate(640.5,0)"
      >
        <line
          stroke="currentColor"
          y2={-6}
        />
        <text
          dy="0em"
          fill="currentColor"
          y={-9}
        >
          Pro-LGBTQ
        </text>
      </g>
    </g>
  </svg>
</div>
`;
